neqo-transport/src/connection.rs

```diff
@@ -3223,4 +3249,214 @@ mod tests {
+        while let Output::Datagram(dg) = server.process_output(now) {
+            s_tx_dgrams.push(dg);
+        }
+
+        // ACK pkts 0-10
+        assert_eq!(s_tx_dgrams.len(), 1);
+
+        // Handle ack, send more
+        qinfo!("client");
+        for s_dgram in s_tx_dgrams {
+            client.process_input(s_dgram, now)
+        }
+
+        assert_eq!(client.stream_send(2, &[0xab; 100_000]).unwrap(), 49152);
+
+        let mut c_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = client.process_output(now) {
+            c_tx_dgrams.push(dg);
+        }
+
+        assert_eq!(c_tx_dgrams.len(), 21);
+
+        // Recv all pkts and ACK
+        qinfo!("server");
+        for c_dgram in c_tx_dgrams {
+            server.process_input(c_dgram, now)
+        }
+        assert_eq!(server.stream_recv(2, &mut srv_buf).unwrap(), (24617, false));
+        let mut s_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = server.process_output(now) {
+            s_tx_dgrams.push(dg);
+        }
+
+        // ACK pkts 0-31
+        assert_eq!(s_tx_dgrams.len(), 1);
+
+        // Handle ack, send more
+        qinfo!("client");
+        for s_dgram in s_tx_dgrams {
+            client.process_input(s_dgram, now)
+        }
+
+        // Cram more into the stream
+        assert_eq!(client.stream_send(2, &[0xab; 100_000]).unwrap(), 36900);
+
+        let mut c_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = client.process_output(now) {
+            c_tx_dgrams.push(dg);
+        }
+
+        assert_eq!(c_tx_dgrams.len(), 41);
+
+        // Packets lost!
+        qinfo!("server");
+        for _ in 0..25 {
+            c_tx_dgrams.remove(10);
+        }
+        for c_dgram in c_tx_dgrams {
+            server.process_input(c_dgram, now)
+        }
+        assert_eq!(server.stream_recv(2, &mut srv_buf).unwrap(), (12005, false));
+        let mut s_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = server.process_output(now) {
+            s_tx_dgrams.push(dg);
+        }
+        assert_eq!(s_tx_dgrams.len(), 1);
+
+        // Handle ack, transition slow start -> cong avoidance due to dropped
+        // pkts
+        qinfo!("client");
+        for s_dgram in s_tx_dgrams {
+            client.process_input(s_dgram, now)
+        }
+        let cong_cwnd = client.loss_recovery.cwnd();
+        assert_eq!(cong_cwnd, 35116);
+
+        // Cram more into the stream, but nothing added
+        assert_eq!(client.stream_send(2, &[0xac; 100_000]).unwrap(), 0);
+
+        let mut c_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = client.process_output(now) {
+            c_tx_dgrams.push(dg);
+        }
+
+        // Constrained by cwnd
+        assert_eq!(c_tx_dgrams.len(), 29);
+        assert_eq!(
+            c_tx_dgrams.iter().map(|dg| dg.len() as u64).sum::<u64>(),
+            cong_cwnd
+        );
+
+        qinfo!("server");
+        for c_dgram in c_tx_dgrams {
```

Comment:
The function here  is called `server_receives_a_ton_of_packets_with_X_bytes_of_data_and_responds_with_an_ACK`.  That can return the ACK-carrying datagram.

Reply:
Currently called the slightly-shorter `server_ack_bytes()` but yeah.